src/main/java/com/cloudbees/hudson/plugins/folder/ChildNameGenerator.java

```diff
@@ -0,0 +1,290 @@
+public abstract class ChildNameGenerator<P extends AbstractFolder<I>, I extends TopLevelItem> {
+    /**
+     * The name of the file that contains the actual name of the child item.
+     */
+    public static final String CHILD_NAME_FILE = "name-utf8.txt";
+
+    private static final Map<Trace,String> idealNames = new WeakHashMap<Trace,String>();
+
+    /**
+     * Work-around helper method to "fix" {@link Item} constructors that have on-disk side-effects and therefore
+     * need {@link Item#getRootDir()} to work during the constructor.
+     * @param project the {@link AbstractFolder}.
+     * @param itemName the name that will be returned by {@link Item#getName()} when the item is constructed. This is
+     *                 the second parameter of {@link AbstractItem#AbstractItem(ItemGroup, String)}. This one would be
+     *                 the one with URL path segment escaping.
+     * @param idealName the original name before whatever URL path segment escaping you applied
+     * @return the {@link Trace} to keep track of when we can remove the memory of the creation process. Please
+     * {@link Trace#close()} the trace after the item is created.
+     */
+    @Nonnull
+    public static final Trace beforeCreateItem(@Nonnull AbstractFolder<?> project,
+                                               @Nonnull String itemName,
+                                               @Nonnull String idealName) {
+        final Trace trace = new Trace(project, itemName);
+        synchronized (idealNames) {
+            idealNames.put(trace, idealName);
+        }
+        return trace;
+    }
+
+    /**
+     * Clean up for a creation {@link Trace}. Not strictly required, but nice implementations will do this via {@link Trace#close()}.
+     * @param trace the trace.
+     */
+    private static final void afterItemCreated(@Nonnull Trace trace) {
+        synchronized (idealNames) {
+            idealNames.remove(trace);
+        }
+    }
+
+    /**
+     * Looks up the {@link Item} to see if we stored the ideal name before invoking the constructor that is having
+     * on-disk side-effects before the object has escaped {@link #beforeCreateItem(AbstractFolder, String, String)}
+     * @param parent
+     * @param item
+     * @return
+     */
+    @CheckForNull
+    protected final String idealNameFromItem(@Nonnull P parent, @Nonnull I item) {
+        String itemName = item.getName();
+        if (itemName == null) return null;
+        synchronized (idealNames) {
+            return idealNames.get(new Trace(parent, itemName));
+        }
+    }
+
+    /**
+     * Infers the {@link Item#getName()} from the {@link Item} instance itself. For a valid implementation, the
+     * {@link ComputedFolder} using this {@link ChildNameGenerator} will be attaching into the {@link Item} the
+     * actual name, typically via a {@link JobProperty} or {@link Action} (beware {@link TransientActionFactory}
```

Comment:
The point of CHILD_NAME_FILE is to allow an admin to figure out which folder is which. It is write-only from our perspective (though we will use it as second last resort in the event that the `Action` or `Property` is not present)